pair/io/src/lib.rs

```diff
@@ -1,229 +1,463 @@
-        amount_out: u128,
     },
+    /// Should be returned from [`InnerAction::Transfer`].
+    Transfer(FTTransfer),
+}
+
+impl From<FTTransfer> for Event {
+    fn from(value: FTTransfer) -> Self {
+        Self::Transfer(value)
+    }
 }
 
-#[derive(Debug, Encode, Decode, TypeInfo)]
-pub enum PairStateQuery {
-    TokenAddresses,
-    Reserves,
-    Prices,
-    BalanceOf(ActorId),
+#[derive(Debug, Encode, Decode, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, TypeInfo, Hash)]
+pub enum SwapKind {
+    AForB,
+    BForA,
 }
 
-#[derive(Debug, Encode, Decode, TypeInfo)]
-pub enum PairStateReply {
-    TokenAddresses((FungibleId, FungibleId)),
-    Reserves((u128, u128)),
-    Prices((u128, u128)),
-    Balance(u128),
+/// Error variants of failed [`Action`].
+#[derive(Debug, Encode, Decode, PartialEq, Eq, PartialOrd, Ord, Clone, TypeInfo, Hash)]
+pub enum Error {
+    /// See [`ContractError`].
+    ContractError(String),
+    /// An insufficient amount of the A or B token was provided.
+    InsufficientAmount,
+    /// A specified amount limit of the former tokens has been exceeded.
+    InsufficientFormerAmount,
+    /// A specified amount limit of the latter tokens has been exceeded.
+    InsufficientLatterAmount,
+    /// An insufficient amount of liquidity tokens was provided, or the contract
+    /// doesn't have enough of them to continue an action.
+    InsufficientLiquidity,
+    /// An invalid recipient was specified.
+    InvalidRecipient,
+    /// [`ActorId::zero()`] was found where it's forbidden.
+    ZeroActorId,
+    /// One of the contract's FT contracts failed to complete a transfer
+    /// action.
+    ///
+    /// Most often, the reason is that a user didn't give an approval to the
+    /// contract or didn't have enough tokens to transfer.
+    TransferFailed,
+    /// An overflow occurred during calculations.
+    Overflow,
+    /// A specified deadline for an action was exceeded.
+    DeadlineExceeded,
+    /// SFT [`ActorId`]s in a given pair to create the Pair contract are equal.
+    IdenticalTokens,
+    FTError(FTError),
+    /// The contract failed to get fee receiver (`fee_to`) [`ActorId`] from the
+    /// linked Factory contract.
+    FeeToGettingFailed,
+    TxCacheError(TransactionManagerError),
+}
+
+impl From<ContractError> for Error {
+    fn from(error: ContractError) -> Self {
+        Self::ContractError(error.to_string())
+    }
+}
+
+impl From<TransactionManagerError> for Error {
+    fn from(error: TransactionManagerError) -> Self {
+        Self::TxCacheError(error)
+    }
+}
+
+impl From<FTError> for Error {
+    fn from(error: FTError) -> Self {
+        Self::FTError(error)
+    }
+}
+
+#[derive(Debug, Encode, Decode, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, TypeInfo, Hash)]
+pub enum CachedAction {
+    Swap(u128),
+    AddLiquidity((u128, u128)),
+    RemovedLiquidity { amount: Amount, is_burned: bool },
+    Other,
+}
+
+#[doc(hidden)]
+pub mod hidden {
```

Comment:
Why is it hidden?

Reply:
It's private shared functions for the `pair` & `pair-state` crates. I though about including them in the `pair` crate, but IMO nothing in a main crate (i.e. `pair`) should be public because that's what `pair-io` is for.